[quant][core][gpu][improvement] Set tensors as virtual in quantized cudnn linear op

dzdang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75969
* __->__ #75967

Summary:
For cudnn, if certain intermediate tensors are not needed for other
parts of the operation graph in cudnn, the it can be set as virtual, so
that a pointer-uid pair does not need to be provided for the variant
pack, allowing for better performance.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35717354](https://our.internmc.facebook.com/intern/diff/D35717354)

dzdang committed:
[quant][core][gpu][improvement] Set tensors as virtual in quantized cudnn linear op

Summary:
For cudnn, if certain intermediate tensors are not needed for other
parts of the operation graph in cudnn, the it can be set as virtual, so
that a pointer-uid pair does not need to be provided for the variant
pack, allowing for better performance.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Set tensors as virtual in quantized cudnn linear op"


Summary:
For cudnn, if certain intermediate tensors are not needed for other
parts of the operation graph in cudnn, the it can be set as virtual, so
that a pointer-uid pair does not need to be provided for the variant
pack, allowing for better performance.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

[ghstack-poisoned]

dzdang committed:
Update on "[quant][core][gpu][improvement] Set tensors as virtual in quantized cudnn linear op"


Summary:
For cudnn, if certain intermediate tensors are not needed for other
parts of the operation graph in cudnn, the it can be set as virtual, so
that a pointer-uid pair does not need to be provided for the variant
pack, allowing for better performance.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35717354).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -147,26 +147,15 @@ void PackedLinearWeightCudnn::apply_impl_helper(const at::Tensor& quantized_outp
+    data_ptrs.reserve(9);
+    uids.reserve(9);
+    data_ptrs = {input_fp.data_ptr(), weight_fp.data_ptr(), requantize_multiplier_tensor.data_ptr(),
+                 reinterpret_cast<int8_t*>(quantized_output.data_ptr())};
```
can we use data_ptr<int8_t> now?

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/Linear.cpp:
```diff
@@ -237,7 +228,7 @@ void PackedLinearWeightCudnn::apply_impl_helper(const at::Tensor& quantized_outp
     relu_op.emplace(cudnn_frontend::OperationBuilder(CUDNN_BACKEND_OPERATION_POINTWISE_DESCRIPTOR)
       .setxDesc(tensor2requant_ptr)
-      .setyDesc(cudnn_utils::getTensorDescriptor(linear_output, 'f', key.output_alignment))
+      .setyDesc(cudnn_utils::getTensorDescriptor(linear_output, 'f', key.output_alignment, true))
```
is this the final output? so setting final output as virtual would speed things up?

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35717354).

jerryzh168 commented on aten/src/ATen/native/quantized/cudnn/utils.h:
```diff
@@ -193,9 +193,19 @@ uint8_t getAlignment(const at::Tensor &t) {
 }
 
-cudnn_frontend::Tensor getTensorDescriptor(const at::Tensor &t, int64_t id, uint8_t alignment) {
+cudnn_frontend::Tensor getTensorDescriptor(const at::Tensor &t, int64_t id, uint8_t alignment, bool is_virtual = false) {
```
please add some docs for `is_virtual` here

dzdang committed:
Update on "[quant][core][gpu][improvement] Set tensors as virtual in quantized cudnn linear op"


Summary:
For cudnn, if certain intermediate tensors are not needed for other
parts of the operation graph in cudnn, the it can be set as virtual, so
that a pointer-uid pair does not need to be provided for the variant
pack, allowing for better performance.

Test plan:
```
python test/test_quantization.py -k test_qlinear_cudnn
```

Differential Revision: [D35717354](https://our.internmc.facebook.com/intern/diff/D35717354)

[ghstack-poisoned]

dzdang commented:
@dzdang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35717354).
